File: src/main/java/io/invest/iagent/service/extraction/extractor/PeriodTypeUtil.java
package io.invest.iagent.service.extraction.extractor;

import com.google.common.collect.Lists;
import io.invest.iagent.service.extraction.model.FinancialTable;
import io.invest.iagent.service.extraction.model.TableRow;
import org.apache.commons.lang3.StringUtils;
import org.springframework.util.CollectionUtils;

import java.util.List;
import java.util.Optional;

public class PeriodTypeUtil {

    public static String determinePeriodType(FinancialTable table){
        String lowerTitle = Optional.ofNullable(table.getTitle()).map(String::toLowerCase).orElse("");
        String periodType = determinePeriodType(lowerTitle);
        // by headers
        if(StringUtils.isBlank(periodType)){
            periodType = Optional.ofNullable(table.getHeaders()).orElse(Lists.newArrayList())
                    .stream().map(PeriodTypeUtil::periodType)
                    .filter(StringUtils::isNotBlank)
                    .findFirst().orElse(null) ;
        }
        // by rows
        if(StringUtils.isBlank(periodType)){
            periodType = periodType(table.getRows()) ;
        }
        return periodType ;
    }


    private static String periodType(List<TableRow> rows){
        if(CollectionUtils.isEmpty(rows)){
            return null ;
        }
        // year
        boolean hasYearEnded = rows.stream()
                .anyMatch(row -> contains(row, "year ended"));
        if(hasYearEnded){
            return "FY" ;
        }
        // half
//        boolean hasHalfEnded = rows.stream()
//                .anyMatch(row -> contains(row, "six months ended"));
//        if(hasHalfEnded){
//            return "H" ;
//        }
        // quarter
        boolean hasQuarterEnded = rows.stream()
                .anyMatch(row -> contains(row, "three months ended"));
        if(hasQuarterEnded){
            if(rows.stream().anyMatch(r->contains(r,"march"))){
                return "Q1" ;
            } else if (rows.stream().anyMatch(r->contains(r,"june"))) {
                return "Q2" ;
            } else if (rows.stream().anyMatch(r->contains(r,"september"))) {
                return "Q3" ;
            } else if (rows.stream().anyMatch(r->contains(r,"december"))) {
                return "Q4" ;
            }
        }
        return null ;
    }

    private static boolean contains(TableRow row, String keyword){
        // label
        String label = Optional.ofNullable(row).map(TableRow::getLabel)
                .map(String::toLowerCase).orElse("") ;
        boolean contains = label.contains(keyword) ;
        if(contains){
            return true ;
        }
        // cells
        contains = Optional.ofNullable(row)
                .map(TableRow::getCells)
                .stream().anyMatch(cells->cells
                        .stream().filter(c-> c!=null && c.getText()!=null)
                        .anyMatch(cell -> cell.getText().toLowerCase().contains(keyword)));
        return contains ;
    }


    /**
     * 根据标题确定周期类型
     * Q1: 1-3月，Q2: 4-6月，Q3: 7-9月，Q4: 10-12月
     * H1: 上半年，H2: 下半年
     * FY: 全年/年报
     */
    private static String determinePeriodType(String lowerTitle) {
        if (lowerTitle == null || lowerTitle.isEmpty()) {
            return "";
        }

        // 判断季度
        if (lowerTitle.contains("three months") || lowerTitle.contains("quarter")) {
            if (lowerTitle.contains("march") || lowerTitle.contains("march 31") ||
                    lowerTitle.contains("february") || lowerTitle.contains("january")) {
                return "Q1";
            } else if (lowerTitle.contains("june") || lowerTitle.contains("may") ||
                    lowerTitle.contains("april")) {
                return "Q2";
            } else if (lowerTitle.contains("september") || lowerTitle.contains("august") ||
                    lowerTitle.contains("july")) {
                return "Q3";
            } else if (lowerTitle.contains("december") || lowerTitle.contains("november") ||
                    lowerTitle.contains("october")) {
                return "Q4";
            }
            // 默认Q1（很多公司财年从3月结束）
            return "Q1";
        }

        // 判断半年报
        if (lowerTitle.contains("six months") || lowerTitle.contains("six-month") ||
                lowerTitle.contains("half year")) {
            if (lowerTitle.contains("june") || lowerTitle.contains("first half")) {
                return "H1";
            }
            return "H2";
        }

        // 判断年报
        if (lowerTitle.contains("year ended") || lowerTitle.contains("fiscal year") ||
                lowerTitle.contains("twelve months") || lowerTitle.contains("12 months") ||
                lowerTitle.contains("full year")) {
            return "FY";
        }

        return "";
    }

    /**
     * 推断周期
     */
    private static String periodType(String header){
        if(StringUtils.isBlank(header)){
            return null ;
        }
        header = header.toLowerCase() ;
        if(header.contains("year ended")){
            return "FY" ;
        }
        if(header.contains("six months ended")){
            return "H" ;
        }
        if(header.contains("three months ended")){
            if(header.contains("march")){
                return "Q1" ;
            }else if(header.contains("june")){
                return "Q2" ;
            } else if (header.contains("september")) {
                return "Q3" ;
            }else if (header.contains("december")) {
                return "Q4";
            }
        }
        return null ;
    }

}

File: workspace/skills/segment-financial-report/scripts/engine/period_type_util.py
# -*- coding: utf-8 -*-
"""表格级周期分类器：判断 FinancialTable 属于 FY/H1/Q1..Q4。

支持通过 ``fiscal_year_end_month`` 指定公司的财年结束月（默认 12 = 日历年），
从而正确将 Mar/Jun/Sep/Dec 等月份名映射到对应财季。例如：
- fiscal_year_end_month=12（默认，日历年）：Mar→Q1, Jun→Q2, Sep→Q3, Dec→Q4
- fiscal_year_end_month=6（Microsoft 等 6 月财年结束公司）：Sep→Q1, Dec→Q2, Mar→Q3, Jun→Q4
- fiscal_year_end_month=3（阿里等零售/HK 3 月财年结束）：Jun→Q1, Sep→Q2, Dec→Q3, Mar→Q4
"""
from __future__ import annotations

import re
from typing import Optional

from .model import FinancialTable, TableRow


# 月份名（1 月起）与 0 基下标
MONTH_LIST = ["january", "february", "march", "april", "may", "june",
              "july", "august", "september", "october", "november", "december"]
_MONTH_INDEX = {name: i for i, name in enumerate(MONTH_LIST)}
# 财历溢出阈值：结束日落在次月 1~7 号视为上一自然月所属财季（见 effective_period_month）
_SPILL_DAY = 7


def effective_period_month(lower_text: Optional[str], month: Optional[str]) -> Optional[str]:
    """4-4-5/13 周财历（如 Apple）财季结束日可能落在次月初 1~7 号
    （如 FY2023 Q2 结束于 "April 1, 2023"、Q3 结束于 "July 1, 2023"）。
    这类结束日经济上属于上一自然月对应的财季；若不归一，April 会被映射成财年 Q3、
    July 成 Q4，导致该季整体错位一季。月末（≥25 号）等正常结束日不受影响。

    :param lower_text: 含 "<Month> <day>" 的期间文本（已转小写）
    :param month:      文本中识别出的月份名
    :return:           归一后用于定财季的月份名
    """
    if not month or not lower_text:
        return month
    rx = re.compile(r"\b" + re.escape(month) + r"\s+(\d{1,2})\b")
    m = rx.search(lower_text)
    if m and int(m.group(1)) <= _SPILL_DAY:
        return MONTH_LIST[(_MONTH_INDEX[month] - 1) % 12]
    return month


# 日历年（Dec FY-end）月份→财季映射
_CALENDAR_MONTH_TO_QUARTER = {
    # Q1
    "january": "Q1", "february": "Q1", "march": "Q1",
    # Q2
    "april": "Q2", "may": "Q2", "june": "Q2",
    # Q3
    "july": "Q3", "august": "Q3", "september": "Q3",
    # Q4
    "october": "Q4", "november": "Q4", "december": "Q4",
}


def _build_month_to_quarter_map(fy_end_month: int) -> dict:
    """根据财年结束月构建 month name → Qn 的映射。"""
    fy_end_month = int(fy_end_month) if fy_end_month else 12
    if fy_end_month < 1 or fy_end_month > 12:
        fy_end_month = 12
    if fy_end_month == 12:
        return dict(_CALENDAR_MONTH_TO_QUARTER)
    # Q1 结束月 = (fy_end + 3 - 1) % 12 + 1  等等，直接枚举更直观
    month_names = ["january", "february", "march", "april", "may", "june",
                   "july", "august", "september", "october", "november", "december"]
    # 季度结束月（按 Q1, Q2, Q3, Q4 顺序）
    # Q1 ends: (fy_end % 12) + 3 months after FY start; simpler enumerate:
    # FY starts month = fy_end_month + 1 (wrapping)
    # Q1: months [start+0..start+2] 三个月，end = start+2
    # start_month (1-indexed, first month of FY): fy_end_month % 12 + 1
    start = (fy_end_month % 12) + 1  # 1-12
    mapping = {}
    for q in range(1, 5):
        # 该季度的三个月（0-indexed offset 0,1,2 from start）
        for offset in range(3):
            m_idx_0 = (start - 1 + (q - 1) * 3 + offset) % 12
            mapping[month_names[m_idx_0]] = f"Q{q}"
    return mapping


def determinePeriodType(table: FinancialTable, fiscal_year_end_month: int = 12) -> str:
    title = (table.getTitle() or "").lower()
    m2q = _build_month_to_quarter_map(fiscal_year_end_month)
    pt = _from_title(title, m2q)
    if not pt:
        # Use document-level period hint (e.g., "Three Months Ended June 30, 2025"
        # captured from the text preceding the table in press-release HTML).
        hint = (table.period or "").lower()
        if hint:
            # _from_title works on any period-bearing text (it checks for
            # three/six/nine/twelve months keywords and month names).
            pt = _from_title(hint, m2q)
            if not pt:
                pt = _from_period_phrase(hint, m2q)
    if not pt:
        for h in table.getHeaders() or []:
            r = _from_header(h, m2q)
            if r:
                pt = r
                break
    if not pt:
        pt = _from_rows(table.getRows(), m2q)
    return pt or ""


def _from_period_phrase(lower_text: str, m2q: dict) -> str:
    """Extract quarter/FY from a phrase like 'quarter ended june 30, 2025'."""
    if not lower_text:
        return ""
    if "year ended" in lower_text or "fiscal year ended" in lower_text or "twelve months ended" in lower_text:
        return "FY"
    if "nine months ended" in lower_text:
        return "QTD9"
    if "six months ended" in lower_text or "half year" in lower_text:
        return "QTD6"
    if "quarter ended" in lower_text or "three months ended" in lower_text:
        mon = effective_period_month(lower_text, _month_from_text(lower_text))
        if mon:
            return m2q.get(mon, "Q1")
        return "Q1"
    return ""


def _from_rows(rows, m2q: dict) -> Optional[str]:
    """Detect the table's primary period type by scanning cells for period keywords.

    When a table mixes quarter columns with YTD/FY columns (common in H1/Q2 press
    releases which show both Three Months and Six Months Ended), prefer the
    single-quarter type so that period_sequence can correctly label each column
    (QTD columns are filtered downstream). FY is detected first as it dominates
    annual reports.
    """
    if not rows:
        return None
    has_fy = any(_row_contains(r, kw) for r in rows
                 for kw in ("year ended", "twelve months ended",
                            "fiscal year ended", "12 months ended"))
    has_qtd9 = any(_row_contains(r, kw) for r in rows
                   for kw in ("nine months ended", "9 months ended"))
    has_qtd6 = any(_row_contains(r, kw) for r in rows
                   for kw in ("six months ended", "6 months ended", "half year"))
    has_quarter = any(_row_contains(r, kw) for r in rows
                      for kw in ("three months ended", "three month ended",
                                 "three-month", "3 months ended", "3 month ended",
                                 "quarter ended"))
    # FY dominates (annual report table, no single-quarter breakdown)
    if has_fy and not has_quarter and not has_qtd9 and not has_qtd6:
        return "FY"
    # Single-quarter takes precedence over YTD for mixed press-release tables
    # (e.g. Q2 releases show both Three Months and Six Months Ended).
    if has_quarter:
        latest = _find_latest_year_month(rows)
        if latest is not None:
            _y, mname = latest
            return m2q.get(mname, "Q1")
        return "Q1"
    if has_qtd9:
        return "QTD9"
    if has_qtd6:
        return "QTD6"
    if has_fy:
        return "FY"
    return None


def _row_contains(row: TableRow, keyword: str) -> bool:
    label = (row.getLabel() or "").lower()
    if keyword in label:
        return True
    for c in row.getCells() or []:
        if c is not None and c.getText() is not None:
            if keyword in c.getText().lower():
                return True
    return False


_MONTH_SET = {"january","february","march","april","may","june",
              "july","august","september","october","november","december"}


def _find_latest_year_month(rows) -> Optional[tuple]:
    """Scan all cells for (year, month_name) pairs and return the latest chronologically.

    Looks for phrases like ``september 30, 2025`` or ``june 2025`` in cell text.
    Returns (year_int, month_name_lower) or None if no date found. 返回的月份名已经过
    财历溢出归一（结束日落在次月 1~7 号时归到上一月），比较仍按真实日期先后。
    """
    import re
    # Match <Month> <day>, <year>（day 必填：无日的纯年月无法判断溢出，回退原月）
    _DATE_RE = re.compile(
        r"(january|february|march|april|may|june|july|august|september|"
        r"october|november|december)\s+(\d{1,2})\s*,?\s*(20\d{2})",
        re.IGNORECASE,
    )
    # best: (year, real_month_index, effective_month)
    best: Optional[tuple] = None
    for row in rows or []:
        # Check label
        for text_src in [(row.getLabel() if row else "")] + [
            (c.getText() if c else "") for c in (row.getCells() if row else [])
        ]:
            if not text_src:
                continue
            low = text_src.lower().replace("\xa0", " ")
            for m in _DATE_RE.finditer(low):
                mon = m.group(1).lower()
                day = int(m.group(2))
                yr = int(m.group(3))
                eff = MONTH_LIST[(_MONTH_INDEX[mon] - 1) % 12] if day <= _SPILL_DAY else mon
                cand = (yr, _MONTH_INDEX[mon], eff)
                if best is None or (cand[0], cand[1]) > (best[0], best[1]):
                    best = cand
    if best is None:
        return None
    return (best[0], best[2])


def _date_tuple_gt(a: tuple, b: tuple) -> bool:
    """Return True if (year, month_name) a is later than b."""
    month_order = {n: i + 1 for i, n in enumerate(
        ["january","february","march","april","may","june",
         "july","august","september","october","november","december"])}
    if a[0] != b[0]:
        return a[0] > b[0]
    return month_order.get(a[1], 0) > month_order.get(b[1], 0)


def _month_from_text(lower_text: str) -> Optional[str]:
    """Return the latest month name appearing in lower_text, or None."""
    found = None
    month_order = {n: i + 1 for i, n in enumerate(
        ["january","february","march","april","may","june",
         "july","august","september","october","november","december"])}
    for mon in _MONTH_SET:
        if mon in lower_text:
            if found is None or month_order[mon] > month_order[found]:
                found = mon
    return found


def _from_title(lower_title: str, m2q: dict) -> str:
    if not lower_title:
        return ""
    if any(x in lower_title for x in (
            "year ended", "fiscal year", "twelve months", "12 months", "full year")):
        return "FY"
    if "nine months" in lower_title or "9 months" in lower_title:
        return "QTD9"
    if any(x in lower_title for x in ("six months", "six-month", "half year")):
        return "QTD6"
    if "three months" in lower_title or "three-month" in lower_title or "quarter" in lower_title:
        mon = effective_period_month(lower_title, _month_from_text(lower_title))
        if mon:
            return m2q.get(mon, "Q1")
        return "Q1"
    return ""


def _from_header(header: Optional[str], m2q: dict) -> Optional[str]:
    if not header or not header.strip():
        return None
    h = header.lower().replace("\xa0", " ")
    if any(x in h for x in ("year ended", "fiscal year ended", "twelve months ended")):
        return "FY"
    if "nine months ended" in h or "9 months ended" in h:
        return "QTD9"
    if "six months ended" in h or "6 months ended" in h or "half year" in h:
        return "QTD6"
    if any(x in h for x in ("three months ended", "three month ended",
                            "three-month", "3 months ended", "quarter ended")):
        mon = effective_period_month(h, _month_from_text(h))
        if mon:
            return m2q.get(mon, "Q1")
        return "Q1"
    return None
